### PR TITLE
Let `prefix_allowed_tokens_fn` override model `-inf` and raise an exception on unsatisfiable generation constraints.

### LINKED ISSUE #22890: `prefix_allowed_tokens_fn` do not constrain when all allowed tokens have scores of `-inf`
### System Info

transformer 4.25.1
python 3.8.16

### Who can help?

@gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

When using `generate()` with `prefix_allowed_tokens_fn`, (more precisely, when using `PrefixConstrainedLogitsProcessor`),
when all tokens returned by `prefix_allowed_tokens_fn` have scores of `-inf`, the model does not comply with the constraints and picks the token which is not on the allowed token list.

### Expected behavior

Even if all allowed tokens have score of `-inf`, the model should pick tokens from allowed token list by `prefix_allowed_tokens_fn`.
I think it can be solved by using some clamp function or adding epsilon value to this code.
https://github.com/huggingface/transformers/blob/04ab5605fbb4ef207b10bf2772d88c53fc242e83/src/transformers/generation/logits_process.py#L692-L698

This is my own code to solve it.
However, it might cause other bugs.

```python
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
        masked_score = torch.full_like(scores, -math.inf)
        for batch_id, beam_sent in enumerate(input_ids.view(-1, self._num_beams, input_ids.shape[-1])):
            for beam_id, sent in enumerate(beam_sent):
                allowed_idx = batch_id * self._num_beams + beam_id, self._prefix_allowed_tokens_fn(batch_id, sent)
                filtered_scores = torch.clamp(scores[allowed_idx], min=-10 ** 6)
                masked_score[allowed_idx] = filtered_scores
        return masked_score
```
Edit: The model works well on `torch.clamp()` with `min=-10 ** 6`, not `min=-10 ** 8`, when all allowed token's score is -inf. Too low score token in the sequence may have affected the decoding step. I updated the above code.

### PR DESCRIPTION
**Pull request by ksh108405**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48927) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48927)

# What does this PR do?

Fixes #22890

When every token allowed by `prefix_allowed_tokens_fn` already has a score of `-inf`, the constraint is silently
dropped.
This PR separates the two possible cause of `-inf` and handles them differently.

- `-inf` produced by the model (e.g. fp16 overflow)
  - The constraint should still be honored.
  - `PrefixConstrainedLogitsProcessor` now runs earlier than the the masking processors. If the allowed tokens of every beam of a batch member are `-inf`, it forces them with a score of `0`, the same convention as `ForcedBOS/EOSTokenLogitsProcessor`.
- `-inf` produced by other processors (e.g. `min_new_tokens`, `bad_words_ids`, ...)
  - The constraints contradict each other. It's reasonable to raise an exception.
  - Since the conflict is only visible on the final distribution, the check had to go into the decoding loops. Now, if `_sample` and `_beam_search` detect rows without any finite candidate, they raise `ValueError` to inform it.

## Who can review?

@Rocketknight1

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
On the conflict path we still select and stream a token before raising. The dead row gets zero scores, `unfinished_sequences` is cleared, then `StopCheck` calls `streamer.put(tokens)`; only after the loop do we raise the constraint error. So an unsatisfiable request can leak a bogus/pad token to a live streamer. Could conflicted rows stop before token selection/streaming, with a streamer regression?

**Review by Rocketknight1 — APPROVED** [review]
I liked the core change, but simplified out the rest! Should still be fine like this, and makes me more confident about the changes
